File: frontend/src/app/services/car.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Car, DashboardStats, CarFilter } from '../models/car.model';

/**
 * Service for car-related API calls
 */
@Injectable({
  providedIn: 'root'
})
export class CarService {
  private apiUrl = 'http://localhost:8080/api';

  constructor(private http: HttpClient) {}

  /**
   * Get all cars
   */
  getAllCars(): Observable<Car[]> {
    return this.http.get<Car[]>(`${this.apiUrl}/cars`);
  }

  /**
   * Get car by ID
   */
  getCarById(id: number): Observable<Car> {
    return this.http.get<Car>(`${this.apiUrl}/cars/${id}`);
  }

  /**
   * Get featured cars
   */
  getFeaturedCars(): Observable<Car[]> {
    return this.http.get<Car[]>(`${this.apiUrl}/cars/featured`);
  }

  /**
   * Filter cars
   */
  filterCars(filter: CarFilter): Observable<Car[]> {
    let params = new HttpParams();
    
    if (filter.type) params = params.set('type', filter.type);
    if (filter.brand) params = params.set('brand', filter.brand);
    if (filter.startYear) params = params.set('startYear', filter.startYear.toString());
    if (filter.endYear) params = params.set('endYear', filter.endYear.toString());
    if (filter.minPrice) params = params.set('minPrice', filter.minPrice.toString());
    if (filter.maxPrice) params = params.set('maxPrice', filter.maxPrice.toString());

    return this.http.get<Car[]>(`${this.apiUrl}/cars/filter`, { params });
  }

  /**
   * Get distinct types
   */
  getTypes(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/cars/types`);
  }

  /**
   * Get distinct brands
   */
  getBrands(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/cars/brands`);
  }

  /**
   * Create a new car
   */
  createCar(car: Car): Observable<Car> {
    return this.http.post<Car>(`${this.apiUrl}/cars`, car);
  }

  /**
   * Update a car
   */
  updateCar(id: number, car: Car): Observable<Car> {
    return this.http.put<Car>(`${this.apiUrl}/cars/${id}`, car);
  }

  /**
   * Delete a car
   */
  deleteCar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/cars/${id}`);
  }

  /**
   * Get dashboard statistics
   */
  getDashboardStats(): Observable<DashboardStats> {
    return this.http.get<DashboardStats>(`${this.apiUrl}/dashboard/stats`);
  }
}

File: backend/src/main/java/com/automobile/app/service/CarService.java
package com.automobile.app.service;

import com.automobile.app.entity.Car;
import com.automobile.app.repository.CarRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

/**
 * Service layer for Car operations
 */
@Service
public class CarService {

    private final CarRepository carRepository;

    public CarService(CarRepository carRepository) {
        this.carRepository = carRepository;
    }

    /**
     * Get all cars
     */
    public List<Car> getAllCars() {
        return carRepository.findAll();
    }

    /**
     * Get car by ID
     */
    public Optional<Car> getCarById(Long id) {
        return carRepository.findById(id);
    }

    /**
     * Create a new car
     */
    @Transactional
    public Car createCar(Car car) {
        return carRepository.save(car);
    }

    /**
     * Update an existing car
     */
    @Transactional
    public Car updateCar(Long id, Car carDetails) {
        Car car = carRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Car not found with id: " + id));
        
        car.setName(carDetails.getName());
        car.setBrand(carDetails.getBrand());
        car.setType(carDetails.getType());
        car.setYear(carDetails.getYear());
        car.setPrice(carDetails.getPrice());
        car.setDescription(carDetails.getDescription());
        car.setImageUrl(carDetails.getImageUrl());
        car.setEngineType(carDetails.getEngineType());
        car.setHorsepower(carDetails.getHorsepower());
        car.setTorque(carDetails.getTorque());
        car.setDisplacement(carDetails.getDisplacement());
        car.setTopSpeed(carDetails.getTopSpeed());
        car.setAcceleration0To100(carDetails.getAcceleration0To100());
        car.setFuelEconomy(carDetails.getFuelEconomy());
        car.setTechFeatures(carDetails.getTechFeatures());
        car.setSafetyRating(carDetails.getSafetyRating());
        car.setPerformanceRating(carDetails.getPerformanceRating());
        car.setIsFeatured(carDetails.getIsFeatured());
        
        return carRepository.save(car);
    }

    /**
     * Delete a car
     */
    @Transactional
    public void deleteCar(Long id) {
        carRepository.deleteById(id);
    }

    /**
     * Get featured cars
     */
    public List<Car> getFeaturedCars() {
        return carRepository.findByIsFeaturedTrue();
    }

    /**
     * Filter cars by multiple criteria
     */
    public List<Car> filterCars(String type, Integer startYear, Integer endYear,
                                BigDecimal minPrice, BigDecimal maxPrice, String brand) {
        if (type != null && startYear != null && endYear != null && 
            minPrice != null && maxPrice != null) {
            return carRepository.findByTypeAndYearBetweenAndPriceBetween(
                    type, startYear, endYear, minPrice, maxPrice);
        }
        
        if (type != null) {
            return carRepository.findByType(type);
        }
        
        if (brand != null) {
            return carRepository.findByBrand(brand);
        }
        
        if (startYear != null && endYear != null) {
            return carRepository.findByYearBetween(startYear, endYear);
        }
        
        if (minPrice != null && maxPrice != null) {
            return carRepository.findByPriceBetween(minPrice, maxPrice);
        }
        
        return carRepository.findAll();
    }

    /**
     * Get distinct types
     */
    public List<String> getDistinctTypes() {
        return carRepository.findDistinctTypes();
    }

    /**
     * Get distinct brands
     */
    public List<String> getDistinctBrands() {
        return carRepository.findDistinctBrands();
    }
}

